lib/sshutils/server.go

```diff
@@ -810,16 +840,57 @@ func (c *ConnectionWrapper) Read(b []byte) (int, error) {
 			skip = payloadBoundary + 1
 		}
 	}
-	c.upstreamReader = io.MultiReader(bytes.NewBuffer(buff[skip:]), c.Conn)
+	if bytes.HasPrefix(buff, multiplexer.ProxyV2Prefix) {
+		reader := bufio.NewReader(io.MultiReader(bytes.NewBuffer(buff), c.Conn))
+
+		proxyLine, err := multiplexer.ReadProxyLineV2(reader)
+		if err != nil {
+			return 0, trace.Wrap(err)
+		}
+		if proxyLine != nil && proxyLine.IsSigned() && c.caGetter != nil {
+			err = proxyLine.VerifySignature(context.Background(), c.caGetter, c.clusterName, c.clock)
```

Comment:
The unlimited duration context makes me a bit scared, not gonna lie. Is there at least some sensible timeout we could have here?

Reply:
I've added 5 seconds timeout 👍